crates/agent/src/lib.rs

```diff
@@ -32,33 +55,107 @@ pub async fn serve<E>(
+    //    that came in while we weren't running
+    // NOTE: it is critical that we use an unbounded channel here, otherwise we would open ourselves up to a deadlock scenario
+    // Example deadlock when using a bounded channel:
+    // 1. A spike of work comes in
+    let (task_tx, mut task_rx) = mpsc::unbounded_channel::<String>();
+
+    // Each handler gets run at least once to check if there is any pending work
+    for (handler_table, _) in handlers_by_table.iter() {
+        task_tx.send(handler_table.to_string())?;
+    }
+
+    let task_tx_cloned = task_tx.clone();
+    let listen_to_datbase_notifications = async move {
+        loop {
+            let item = listener.recv().await.map_err(|e| anyhow::Error::from(e))?;
+            let notification: AgentNotification = serde_json::from_str(item.payload())
+                .context("deserializing agent task notification")?;
+            let table: &str = &notification.table;
+
+            tracing::debug!(table = table, "Message received to invoke handler");
+            task_tx_cloned
+                .send(table.to_string())
+                .map_err(|e| anyhow::Error::from(e))?
+        }
+        // Need this here to indicate that this future returns an anyhow::Result<()>
+        #[allow(unreachable_code)]
+        Ok::<(), anyhow::Error>(())
+    };
+
+    let mut listener_handler = tokio::spawn(listen_to_datbase_notifications);
 
     tokio::pin!(exit);
+
     loop {
-        // Pick handler with the next deadline.
-        let (handler, deadline) = handlers
-            .iter_mut()
-            .min_by_key(|i| i.1)
-            .expect("handlers is not empty");
-
-        // Sleep until its deadline has elapsed.
-        let sleep = tokio::time::sleep_until(*deadline);
         tokio::select! {
             _ = &mut exit => {
                 tracing::debug!("caught signal; exiting...");
                 return Ok(()) // All done.
             }
-            _ = sleep => (),
-        };
+            listener_res = &mut listener_handler => {
+                match listener_res {
+                    // It should be impossible to get here since `listen_to_datbase_notifications` loop has no `return`s or `break`s
+                    Ok(Ok(())) => return Err(anyhow::anyhow!("Unexpected notification listener exit")),
+                    // If we get an error from inside `listen_to_datbase_notifications`,
+                    // something went wrong when actually listening to the postgres channel
+                    Ok(Err(e)) => return Err(e.into()),
+                    // If we got a JoinError, the listener failed unexpectedly
+                    Err(e) => return Err(e.into()),
+                }
+            }
+            maybe_handler_table = task_rx.recv() => {
+                match maybe_handler_table {
+                    Some(handler_table) => {
+                        let mut handler = match handlers_by_table.get(&handler_table as &str) {
```

Comment:
nit: the cyclomatic complexity here is getting pretty gnarly. How about having `tokio::select` emit a `handler_table` which is further handled outside of its body?

The new `let ... else` syntax might also help in successive the destructuring here